### DISCUSSION TITLE
Transparent Images (PNG) has black background when converted.

### DISCUSSION THREAD

**Discussion by ghost** [spatie/laravel-medialibrary#3839] [Q&A]
As title says, when a conversion is made on transparent png image, the converted image has black background.

The new version of the package started having this issue, installed version: `11.0.4`
There was an additional issue with `Fit::Crop` which has been resolved by installing `spatie/images: 3.3.3`.

```php
$this->addMediaConversion('square-thumb')
    ->fit(Fit::Crop, 100, 100)
    ->keepOriginalImageFormat()
    ->performOnCollections('image');
```

Original Image:
![compare-model-3-performance](https://github.com/spatie/laravel-medialibrary/assets/138443677/9896d2a5-2731-42a5-990b-9546388dd4dc)

Converted Image:
![compare-model-3-performance-square-thumb](https://github.com/spatie/laravel-medialibrary/assets/138443677/ea5044ef-d600-4b08-ba31-6305b3db9e96)

**Comment by ghost** [comment]
@timvandijck  Any update regarding this fix?

**Comment by langeuh** [comment]
Having the same problem since v11
Note that this only happens for me when using 'gd' as IMAGE_DRIVER
'imagick' renders the background transparent for the same file.

Don't know if related, but I'm also getting this error for my transparent files:

> ErrorException imagecreatefromstring(): gd-png: libpng warning: iCCP: known incorrect sRGB profile
    vendor/spatie/image/src/Drivers/Gd/GdDriver.php:88 Illuminate\Foundation\Bootstrap\HandleExceptions::handleError
    vendor/laravel/framework/src/Illuminate/Foundation/Bootstrap/HandleExceptions.php:255 Illuminate\Foundation\Bootstrap\HandleExceptions::Illuminate\Foundation\Bootstrap\{closure}
    [internal] imagecreatefromstring
    vendor/spatie/image/src/Drivers/Gd/GdDriver.php:88 Spatie\Image\Drivers\Gd\GdDriver::loadFile
    vendor/spatie/image/src/Image.php:49 Spatie\Image\Image::loadFile...

Versions
> php -i | grep 'GD\|ImageMagick'
GD Support => enabled
GD Version => bundled (2.1.0 compatible)
Imagick compiled with ImageMagick version => ImageMagick 7.1.0-52 Q16-HDRI x86_64 36d3408b5:20221106 https://imagemagick.org
Imagick using ImageMagick library version => ImageMagick 7.1.0-52 Q16-HDRI x86_64 36d3408b5:20221106 https://imagemagick.org
ImageMagick copyright => (C) 1999 ImageMagick Studio LLC
ImageMagick release date => 2022-11-06

spatie/image: 3.3.4
spatie/laravel-medialibrary: 11.0.5

**Comment by ghost** [comment]
@langeuh So using `imagick` as Image Driver solves this bug? Been waiting too long. I need this to be working asap. So can you please confirm.

Edit/Update: 
Yes the 'imagick' driver has no issue with transparent images.

**Comment by timvandijck** [comment]
This should be fixed in the latest version of the spatie/image package that is used by the medialibrary package.

**Comment by ghost** [comment]
@timvandijck I checked but still I am getting the black background with the latest spatie/image package. I was initially using `webp` and the result is same. Tried with pngs to, and the result are with black background.

Original
![image](https://github.com/spatie/laravel-medialibrary/assets/138443677/d28b45ea-6618-422f-938f-a465ed3c5697)

Converted:
![image](https://github.com/spatie/laravel-medialibrary/assets/138443677/a0c8e97a-6213-4afc-93a2-e9e21a0431a3)

Package Versions:
<img width="764" alt="image" src="https://github.com/spatie/laravel-medialibrary/assets/138443677/70e45b22-f301-4f72-8846-ded1fa27e056">

**Comment by ghost** [comment]
I checked with `imagick` driver and its working correctly for pngs, but with webp, its the same issue.

**Comment by timvandijck** [comment]
@anuzpandey5 intresting, would you mind running the test in the spatie/image package?

```bash
vendor/bin/pest --filter "works with transparent pngs" 
```

I'm curious if our differences in results are machine specific or if something else is causing it.

**Comment by zub0r** [comment]
@timvandijck This is still an issue.

Tested on the above image with "spatie/laravel-medialibrary": "11.9.1",
happens for both gd and imagick drivers, (ImageMagick 6.9.11-60, GD library Version: 2.3.3)

**Comment by zub0r** [comment]
Ok, so for anyone having this issue after migrating from older media-library - it looks like before the white backgound was the default, now you need to explicitly set in in your conversion with 
`$this->addMediaConversion('some-conversion')->background('#ffffff')`...

**Comment by timvandijck** [comment]
Indeed, I believe in hindsight it was not ideal to have this default changed, but reverting it back might again cause problems for people who expect it to be there.

For now passing any color that includes transparency like `#ffffff` or `rgba(0, 0, 0, 0)` should do the trick.

**Comment by freekmurze** [comment]
Let's fix this when we tag a new major version :thumbnail

**Reply by TonsiTT** [reply]
@freekmurze could we have a look at this? it was converted from issue to discussion

**Comment by mauricekindermann** [comment]
Yup still a problem. For now I have hard coded `background->(#ffffff)`. I have about 500k images/conversions, so regenerating them is not trivial.

**Comment by TonsiTT** [comment]
Is this still planned to be fixed? The issue still seems to be there

**Comment by timvandijck** [comment]
@TonsiTT like said above we will likely fix this when we tag a new major. For now you can use the workaround of passing a transparent background color: 

```php
$this->addMediaConversion('my-conversion')->background('#ffffff');
```

**Comment by TonsiTT** [comment]
> [@TonsiTT](https://github.com/TonsiTT) like said above we will likely fix this when we tag a new major. For now you can use the workaround of passing a transparent background color:
> 
> $this->addMediaConversion('my-conversion')->background('#ffffff');

Thanks, this issue was closed so i assumed this fell between the cracks.
Also, sorry if i am not understanding this correctly, but using ffffff just adds a white background to my (previously) transparent images. So it doesnt really solve the issue unless i am doing something wrong here

**Comment by timvandijck** [comment]
@TonsiTT can you try passing `rgba(0, 0, 0, 0)` or `#fff0`? I might have been wrong with the `#ffffff`. Basically you just need to pass a color with full transparancy.

**Reply by mrleblanc101** [reply]
`->background('#fff0')` is actually yellow, because it's not a valid HEX, not transparent.
Valid HEX are 3 `#fff`, 6 `#ffffff` or 8 `#ffffff00` character long, never 4.
`#ffffff00` would be transparent, but I get an error `color format not supported`.

**Reply by mrleblanc101** [reply]
I just realized that our app was converting the original `png` image to `jpg`, even before upgrading to the latest version.
We never noticed the problem, because in previous version of the plugin, the image background was white, but now it's black.
Since the image was displayed inside a PDF, we never noticed the lost of transparency.

We were trying to fix the wrong problem, 
Using `->background('rgba(0,0,0,0)')` would never have fixed the issue, because `jpg` have no transparency.
We needed to add `keepOriginalImageFormat()` or `format('png')` and now everything works.

**Reply by TonsiTT** [reply]
This does fixes it more or less. hopefully it can be fixed internally in the code so keepOriginalImageFormat() or format('png') isnt needed

**Comment by nodeandbyte** [comment]
It worked well with RGBA colors, thanks for the tips!

```php
$this
    ->addMediaConversion('thumb')
    ->background('rgba(0, 0, 0, 0)')
    ->keepOriginalImageFormat();
```

**Comment by mauricekindermann** [comment]
Nice, I might try that. Just be care of adding keepOriginalImageFormat if you already have conversions, because I did that and all existing conversions stopped working (since by default this package expects all conversions to be JPG, so it always uses a JPG, and if you specify this it will always return .png or .gif etc). So you just need to regenerate your existing conversions.

**Comment by TonsiTT** [comment]
doesnt seem to work for me. my orignal webp image with transparent background become black background. using 'gd' driver
```
$this
    ->addMediaConversion('thumb')
    ->background('rgba(0, 0, 0, 0)')
    ->keepOriginalImageFormat();
```

**Reply by TonsiTT** [reply]
In fact i notice the lqip (data:image/) also has a black background
